feat(errors): add SARIF report generation

royalpinto007 opened the pull request:
### PR Description:  
This PR adds SARIF (Static Analysis Results Interchange Format) report generation functionality to the DMD compiler using the `-verror-style=sarif` option. When this option is passed, the error messages are generated in SARIF format, providing a structured JSON output that can be integrated with code analysis tools, such as GitHub's SARIF viewer.

#### Example:  
Consider the following code in `test.d`:

```d
void main() {
    x = 5;  // Undefined variable 'x' will trigger an error
}
```

When this file is compiled with the SARIF message style using:

```
dmd test.d -verror-style=sarif
```

The output will be generated in SARIF format, as shown below:

```
test.d(2): Error: undefined identifier `x`
{
  "invocation": {
    "executionSuccessful": false
  },
  "results": [
    {
      "location": {
        "artifactLocation": {
          "uri": "file:///path/to/test.d"
        },
        "region": {
          "startLine": 2,
          "startColumn": 5
        }
      },
      "message": "undefined identifier `x`",
      "ruleId": "DMD"
    }
  ],
  "tool": {
    "name": "DMD",
    "version": "v2.110.0"
  }
}
```

This standardized SARIF report format makes it easier to integrate with static analysis tools, ensuring better error reporting and analysis workflows.

dkorpel commented:
Interesting idea, I've never heard of SARIF before. What D compatible tools currently use it?

New compiler flags are generally to be avoided, but we have extensible `ErrorSink` / `DiagnosticReporter` classes for a reason. I think this is a great usecase for dmd as a library, what do you think @RazvanN7 ?

thewilsonator commented on compiler/src/dmd/errors.d:
```diff
@@ -35,6 +35,133 @@ enum ErrorKind
+void generateSarifReport(const ref Loc loc, const(char)* format, va_list ap, ErrorKind kind) nothrow
+{
+    string toolVersion = global.versionString();  // Retrieve dynamic version
+    ToolInformation tool = ToolInformation("DMD", toolVersion);
```
This variable is unused, and causes the buildkite CI Style check to fail

royalpinto007 replied:
Got it! I’ll remove the unused variables and rebase the branch as suggested. I’ll also update `frontend.h` after finishing the current changes. Thanks for pointing it out!

thewilsonator reviewed (commented):
Some non-specific comments:

Rebasing this branch against current `dlang/master` should fix the `pre-commit` failure.

You will also need to update `frontend.h`, see the CircleCI failure, though feel free to leave this until after you are done making changes.

thewilsonator commented on compiler/src/dmd/errors.d:
```diff
@@ -35,6 +35,133 @@ enum ErrorKind
+}
+
+// Struct for Physical Location of the Error
+struct PhysicalLocation {
```
This looks very similar to https://github.com/dlang/dmd/blob/master/compiler/src/dmd/location.d#L235

It would probably be best if you just did a function that takes a `SourceLoc ` and returns a JSON string representation for it.

royalpinto007 replied:
Thanks! I’ll switch to using `OutBuffer` with `ExtractSlice` and create a stateful function for `SourceLoc` to return a JSON string, similar to `location.d`.

thewilsonator replied:
(I think you resolved the wrong conversation here)

royalpinto007 replied:
Sorry for that. Please review- https://github.com/dlang/dmd/pull/16986/commits/d8d93295cc5ad5bcbff0a33c9ee9d6940bf7bb38

thewilsonator commented on compiler/src/dmd/globals.d:
```diff
@@ -170,6 +170,7 @@ extern (C++) struct Param
     bool addMain;           // add a default main() function
     bool allInst;           // generate code for all template instantiations
     bool bitfields;         // support C style bit fields
+    bool sarif = false;     // track if the --sarif flag is passed
```
if you are adding variables here, you will need to update the corresponding header, `globals.h`

royalpinto007 replied:
Sure, but @dkorpel suggested avoiding compiler flags. If that's the approach I should take, then `globals.h` might not need changes either. Please let me know which direction I should move forward with

thewilsonator replied:
Since it seems to work with output on `stderr`, it would probably be most useful if you can manage to make SARIF a [`MessageStyle`](https://github.com/dlang/dmd/blob/master/compiler/src/dmd/location.d#L27), since then you can use `-verror-style=[digitalmars|gnu|sarif]` and output all the compiler output as JSON.

This might be complicated by `pragma(msg)` a bit (the output is printed on stdout), but that can come later.

royalpinto007 replied:
done- https://github.com/dlang/dmd/pull/16986/commits/19f15d6e8712693be161eb82d96bb542d63c2948

thewilsonator commented on compiler/src/dmd/errors.d:
```diff
@@ -457,8 +584,17 @@ extern (C++) void verrorReport(const ref Loc loc, const(char)* format, va_list a
+                generateSarifReport(info.loc, format, ap, info.kind);
+            }
             if (global.params.v.errorLimit && global.errors >= global.params.v.errorLimit)
+            {
+                fprintf(stderr, "error limit (%d) reached, use `-verrors=0` to show all\n", global.params.v.errorLimit);
```
why is this change made?

royalpinto007 replied:
This was added to prevent the compiler from flooding the user with too many errors in case of multiple issues in the same file. For example, if a file has many undefined variables or other repetitive errors, the compiler will stop reporting after a set limit (e.g., 5 errors) and prompt the user to increase the limit with the `-verrors=0` option if they wish to see all errors.

If this feature is not needed, I can remove it to ensure that all errors are displayed without any restriction (like before)

thewilsonator replied:
> If this feature is not needed, I can remove it to ensure that all errors are displayed without any restriction (like before)

Given SARIF is intended to be machine read, I think this is a good idea.

thewilsonator reviewed (commented):
It would be better for all of your use of `[f,sn]printf` to print to an [`OutBuffer`](https://github.com/dlang/dmd/blob/master/compiler/src/dmd/common/outbuffer.d#L39) and then use [`ExtractSlice`](https://github.com/dlang/dmd/blob/master/compiler/src/dmd/common/outbuffer.d#L685C23-L685C35) to get the printed string.

royalpinto007 commented:
> New compiler flags are generally to be avoided, but we have extensible `ErrorSink` / `DiagnosticReporter` classes for a reason. I think this is a great usecase for dmd as a library, what do you think @RazvanN7 ?

Thanks for the input! I’m looking into using `ErrorSink` or `DiagnosticReporter` to generate SARIF output, which could fit well with DMD as a library. (This is part of Milestone 1 under SAoC 24)

ibuclaw commented on compiler/src/dmd/errors.d:
```diff
@@ -483,6 +619,12 @@ extern (C++) void verrorReport(const ref Loc loc, const(char)* format, va_list a
                     verrorPrint(format, ap, info);
+
+                    // Hook: Generate SARIF report only if the --sarif flag is set
+                    if (global.params.sarif)
+                    {
+                        generateSarifReport(info.loc, format, ap, info.kind);
+                    }
```
Has this been checked against sarif readers such as github? What if there are multiple warnings or errors - is it happy with multiple json blocks or do all diagnostics need collating into one?

royalpinto007 replied:
Thanks for the feedback! 

I've noted this and will work on ensuring the SARIF output is properly collated for multiple diagnostics.

royalpinto007 replied:
I'll address handling multiple diagnostics in a separate PR, if that's okay. What do you think, @thewilsonator?

thewilsonator reviewed (commented):
Now `pre-commit` fails because you have some trailing whitespace.

 Please also take a look at [`json.d`](https://github.com/dlang/dmd/blob/master/compiler/src/dmd/json.d) and how it does 
`item`, `arrayStart` etc.

thewilsonator commented on compiler/src/dmd/location.d:
```diff
@@ -28,6 +28,7 @@ enum MessageStyle : ubyte
 {
     digitalmars,  /// filename.d(line): message
     gnu,          /// filename.d:line: message, see https://www.gnu.org/prep/standards/html_node/Errors.html
+    sarif         /// JSON SARIF output
```
It would be nice to link to the documentation for SARIF here like the GNU standards link above

royalpinto007 replied:
Sure, will do!

royalpinto007 replied:
done- https://github.com/dlang/dmd/pull/16986/commits/db26e1813ad79be34514c1be562c8f12afcbc1e6

royalpinto007 commented:
> And i believe it should be the global path to the file
> I got bitten by this when working on my own LSP just yesterday

Hi @ryuukk, 

I have updated (https://github.com/dlang/dmd/pull/16986/commits/c21e244787dfc53dff58bc19e49a8995c441b93a) the report to use the absolute file path in the uri. The FileName.canonicalName method now generates the correct global path, and the output follows the "file://<absolute-path>" format, as per general SARIF specifications.

Example output now:

```
"uri": "file:///home/royalpinto007/d-build-source/test.d"
```

thewilsonator commented on compiler/src/dmd/location.d:
```diff
@@ -28,6 +28,7 @@ enum MessageStyle : ubyte
 {
     digitalmars,  /// filename.d(line): message
     gnu,          /// filename.d:line: message, see https://www.gnu.org/prep/standards/html_node/Errors.html
+    sarif         /// JSON SARIF output, see https://docs.oasis-open.org/sarif/sarif/v2.1.0/sarif-v2.1.0.html
```
You will need to update both `globals.h` and `frontend.h` with this change.

royalpinto007 replied:
Sure, will do.

royalpinto007 replied:
done- https://github.com/dlang/dmd/pull/16986/commits/1205300547f1d478434f00489d0c10d4f5d3ed6f

thewilsonator commented on compiler/src/dmd/mars.d:
```diff
@@ -1034,6 +1034,9 @@ bool parseCommandLine(const ref Strings arguments, const size_t argc, ref Param
+                params.v.messageStyle = MessageStyle.sarif;
+                break;
             default:
                 error("unknown error style '%.*s', must be 'digitalmars' or 'gnu'", cast(int) style.length, style.ptr);
```
also update the error message here

royalpinto007 replied:
done- https://github.com/dlang/dmd/pull/16986/commits/857749b98ed78340c447b682d450a2cb8fcfc2b0

thewilsonator reviewed (commented):
For the `MessageStyle` addition please also update the [CLI](https://github.com/dlang/dmd/blob/master/compiler/src/dmd/cli.d#L828-L834)

You may want to open a new PR with just the `location.d`, `mars.d`, `cli.d`, globals.h` and `frontend.h` changes. Once you have a PR you have authored merged, the CI will run without waiting for approval from reviewers.

thewilsonator commented on compiler/src/dmd/errors.d:
```diff
@@ -35,6 +36,152 @@ enum ErrorKind
+        }
+    }`);
+
+    return result.extractSlice().dup;
```
pretty sure you don't need the `.dup` here. `OutBuffer` allocates the data, and `extractSlice` returns that and resets the buffer.

royalpinto007 replied:
my bad, corrected- https://github.com/dlang/dmd/pull/16986/commits/f1b76bb2b676ee3045055351bda594159db3ab41

royalpinto007 commented:
> You may want to open a new PR with just the `location.d`, `mars.d`, `cli.d`, globals.h`and`frontend.h` changes. Once you have a PR you have authored merged, the CI will run without waiting for approval from reviewers.

For the MessageStyle changes, I will create a separate PR as requested. For now, I am proceeding with the current PR including a compiler flag, as it seems to be the only viable option at this point. Once this PR is merged, I'll follow up with the new PR for the MessageStyle addition and flag removal

thewilsonator reviewed (changes requested):
needs tests

royalpinto007 committed:
feat(errors): add SARIF report generation

Signed-off-by: royalpinto007 <royalpinto007@gmail.com>

fix: remove unwanted comment

Signed-off-by: royalpinto007 <royalpinto007@gmail.com>

test(fail_compilation): add SARIF integration test case

Signed-off-by: royalpinto007 <royalpinto007@gmail.com>

test(fail_compilation): update tests to pass ci

Signed-off-by: royalpinto007 <royalpinto007@gmail.com>

test(fail_compilation): update regex

Signed-off-by: royalpinto007 <royalpinto007@gmail.com>

test(fail_compilation): update tests to pass ci

Signed-off-by: royalpinto007 <royalpinto007@gmail.com>

test(fail_compilation): update tests with regex

Signed-off-by: royalpinto007 <royalpinto007@gmail.com>

test(fail_compilation): update tests with regex

Signed-off-by: royalpinto007 <royalpinto007@gmail.com>

test(fail_compilation): changes in regex based on ci logs

Signed-off-by: royalpinto007 <royalpinto007@gmail.com>

test(fail_compilation): changes in regex based on ci logs

Signed-off-by: royalpinto007 <royalpinto007@gmail.com>

fix(remove): tool version and change uri to relative path

Signed-off-by: royalpinto007 <royalpinto007@gmail.com>

fix(remove): tool version variable

Signed-off-by: royalpinto007 <royalpinto007@gmail.com>

fix(remove): file:// path

Signed-off-by: royalpinto007 <royalpinto007@gmail.com>

thewilsonator commented:
also needs a rebase

thewilsonator commented:
I think you slightly messed up the rebase. The commit message is now that of  #17001

royalpinto007 commented:
> I think you slightly messed up the rebase. The commit message is now that of #17001

fixed, will now work on adding tests

royalpinto007 commented:
> needs tests

I'm trying to create a SARIF test case, but I'm only getting the default error message instead of the SARIF JSON output. Here's the link to the markdown with details: https://github.com/royalpinto007/d-drafts/blob/main/sarif-tests.md.
Could you please guide me on what I might be missing?

maxhaton commented:
> New compiler flags are generally to be avoided, 

Compiler flags that effect behaviour are the devil, a blanket rule is quite stupid and not what I think Walter meant.

> but we have extensible ErrorSink / DiagnosticReporter classes for a reason. 

Yes. Quite.

> I think this is a great usecase for dmd as a library, what do you think @RazvanN7 ?

I don't follow this, are you saying this should be a separate tool or a plugin or something? Either way this would be a massive mistake strategically.

maxhaton commented on compiler/src/dmd/errors.d:
```diff
@@ -473,6 +620,14 @@ extern (C++) void verrorReport(const SourceLoc loc, const(char)* format, va_list
+
+            // Fix: Convert filename to const(char)* using .ptr
+            Loc sourceLoc = Loc(info.loc.filename.ptr, info.loc.linnum, info.loc.charnum);
+            // Hook: Generate SARIF report only if the --sarif flag is set
```
```suggestion

```

don't think we need this. The if statement tells me what I need to know.

royalpinto007 replied:
done

maxhaton commented:
What happens if there is more than one error at the moment? Are they merged into one report?

royalpinto007 commented:
> What happens if there is more than one error at the moment? Are they merged into one report?

As of now, no. I was thinking of having a separate PR to handle that once we get this merged, and also having a `MessageStyle` instead of a compiler flag- https://github.com/dlang/dmd/pull/16986#discussion_r1798805859

thewilsonator commented on compiler/test/fail_compilation/sarif_test.d:
```diff
@@ -0,0 +1,36 @@
+/*
+TEST_OUTPUT:
+---
+fail_compilation/sarif_test.d(35): Error: undefined identifier `x`
```
Part of the reason I suggest to use a message format (or I suppose an [Output](https://github.com/dlang/dmd/blob/master/compiler/src/dmd/globals.d#L85) is so the entire output is valid JSON, but as you said this can be don in a followup PR

thewilsonator replied:
note 
```
FAILED targets:
- fail_compilation/sarif_test.d
```
```diff
     {
       "location": {
         "artifactLocation": {
-          "uri": "file:///home/royalpinto007/d-build-source/dmd/compiler/test/fail_compilation/sarif_test.d"
+          "uri": "file:///home/circleci/dmd/compiler/test/fail_compilation/sarif_test.d"
         },
         "region": {
           "startLine": 35,
@@ -20,6 +20,6 @@ fail_compilation/sarif_test.d(35): Error
   ],
   "tool": {
     "name": "DMD",
-    "version": "v2.110.0-beta.1-324-gab8582b70f-dirty"
+    "version": "v2.110.0-beta.1-336-g311f34671f-dirty"
   }
 }
```
I forget how DMD's test suite matches variable output. @RazvanN7 ?

royalpinto007 replied:
> note

I believe the regex should fix the CI, as mentioned in the test docs, though it fails locally.
`"uri": "$p:(.*[\/\\])?sarif_test\.d$"`
`"version": "$r:v2\..*$"`

RazvanN7 replied:
What is the expected output (the result of the regex) and the actual output (what running the test actually yields) on your machine?

royalpinto007 replied:
> What is the expected output (the result of the regex) and the actual output (what running the test actually yields) on your machine?

This isn't working as expected. Previously, on my local machine, I was getting:  
```
"uri": "file:///home/royalpinto007/d-build-source/dmd/compiler/test/fail_compilation/sarif_test.d"
```

But in the CI logs, it shows:  
```
"uri": "file:///home/circleci/dmd/compiler/test/fail_compilation/sarif_test.d"
```

Similarly, for the version:  
```
Local: "version": "v2.110.0-beta.1-324-gab8582b70f-dirty"  
CI: "version": "v2.110.0-beta.1-336-g311f34671f-dirty"
```

I need to ensure it matches both environments—local and CI (so I need to update the tests, somehow). Regarding the regex, I attempted it following the test docs, but it's not behaving as expected and seems to yield completely random results.

RazvanN7 replied:
The 'uri' seems to be correct, isn't it? It looks like it puts the correct path. As for the version, it seems to me like there isn't really any way to check if the version is correct so you might as well just match on whatever is being printed there. It would help if you could provide a sample of what the failure looks like on your machine (that is what I asked in my previous comment) .

royalpinto007 replied:
I think things are getting a bit complicated and causing some confusion here. Let me simplify it:  

Currently, I have a test file:
https://github.com/dlang/dmd/pull/16986/files#diff-fb7b99e2941c96abf3738a692a33b3fed7048848289f8611ce45a03142400dc0  

It works as expected on my local machine, but it fails in the CI. The CI expects this output:  
```
"uri": "file:///home/circleci/dmd/compiler/test/fail_compilation/sarif_test.d"
"version": "v2.110.0-beta.1-336-g311f34671f-dirty"
```
Instead, my local machine gives the following output:
```
"uri": "file:///home/royalpinto007/d-build-source/dmd/compiler/test/fail_compilation/sarif_test.d"
"version": "v2.110.0-beta.1-324-gab8582b70f-dirty"
```

To address this, I tried following the [docs](https://github.com/dlang/dmd/tree/master/compiler/test#test_output), which suggest making adjustments in the `TEST_OUTPUT`. 

The relevant part states:
---

**TEST_OUTPUT**  
- Newlines get unified for consistency across platforms.  
- The DMD version string (e.g., `v2.084.0-255-g86b608a15-dirty DEBUG`) should be stripped.  
- Paths to test results are replaced with `{{RESULTS_DIR}}`.  
- These special sequences are offered:  
  - **$n$**: Arbitrary number of digits  
  - **$p:<tail>$**: Matches paths ending with `<tail>`  
  - **$?:<choices>$**: Conditional selections based on environment (e.g., OS, model)  
  - **$r:<regex>$**: Regex matching for dynamic output  
---

However, these adjustments didn’t work for me. I need help figuring out how to get consistent output both locally and in the CI. Could you please help me with this?

RazvanN7 replied:
Can you paste the entire test file that is using the regex and the expected + actual output? Right now, the [link that you provided](https://github.com/dlang/dmd/pull/16986/files#diff-fb7b99e2941c96abf3738a692a33b3fed7048848289f8611ce45a03142400dc0) doesn't use any regexes , rather it has the hardcoded path. This obviously fails the testing pipeline. I want to see the output file that you used **that contains the regex** i.e. what the regex is matching on and what the expected output is. I have not seen any test file that uses the regex, therefore I cannot understand why it might fail.

RazvanN7 replied:
You said that this is not working as expected:

```
"uri": "$p:(.*[\/\\])?sarif_test\.d$"
"version": "$r:v2\..*$"
```

I want to see what the tester outputs when you run the test with the test file containing the above regex. It should give you the pair of expected and actual outputs. That way we can understand what's going wrong with the regex.

royalpinto007 replied:
I have also updated the code here. The CI failures can be reviewed once the checks are complete- https://github.com/dlang/dmd/pull/16986/commits/6263c148b0c4ec5a0d7d800370750707f1de1e05

RazvanN7 commented:
>I think this is a great usecase for dmd as a library, what do you think @RazvanN7 ?

> I don't follow this, are you saying this should be a separate tool or a plugin or something? Either way this would be a massive mistake strategically.

It can definitely be implemented using dmd as a library, however, I have no idea which path is better. Is SARIF that popular? I mean, we don't want to end up implementing all the various error reporting mechanisms. @maxhaton why are you saying that it would be a massive strategic mistake to not implement this in the compiler?

Also, I was under the impression that the point of the SAoC project is to enhance dmd's error reporting mechanism by adding some sort of storage for past errors that can be queried (a list/array of errors), however, that has now turned into implementing a new format for error reporting.

royalpinto007 commented:
I changed the path to relative and removed the tool version for now.

Variable outputs don't seem to work with regexes, and I didn't get much help to proceed. I believe I can now move forward with the `MessageStyle` and then tackle other tasks once this gets merged.

maxhaton commented:
> why are you saying that it would be a massive strategic mistake to not implement this in the compiler?

It's a very thin JSON standard for transporting error messages (in effect), why would you not just put that in the compiler itself? This is what gcc, clang, dotnet etc all do

Also consider what "use dmd as a library" as a generic principle means for something that happens after compilation (or after compilation has failed, rather): One now has to mimic, exactly, the way dmd invokes the frontend. This is non-trivial. Let's suppose this has been achieved and that we can actually invoke it (this is non-trivial sometimes) , there's a much more important issue - we have to compile twice!

If dmd has gained a plugin system while I wasn't looking then it would be more tempting but still KISS.

As for why SARIF, Iain suggested and it seemed good onboarding so I put it first on the list so @royalpinto007 could gain some familiarity with the codebase. The rest of the work will be as specified in the project spec depending on how feasible etc etc

WebFreak001 commented:
Awesome! Thanks for SARIF support, that is going to make GitHub projects soooo much better for reviewing PRs with compiler errors and warnings as well as maintenance of branches.
